golem-worker-executor-base/src/invocation.rs

```diff
@@ -407,6 +425,123 @@ async fn invoke<Ctx: WorkerCtx>(
+            .data()
+            .get_public_state()
+            .event_service()
+            .emit_invocation_start(
+                raw_function_name,
+                idempotency_key,
+                store_context.data().is_live(),
+            );
+    }
+
+    let (_, mut task_exits) = {
+        let hyper_request =
+            virtual_export_compat::http_incoming_handler::input_to_hyper_request(function_input)?;
+        let scheme = wasi_http_scheme_from_request(&hyper_request)?;
+        let incoming = store_context
+            .data_mut()
+            .durable_ctx_mut()
+            .as_wasi_http_view()
+            .new_incoming_request(scheme, hyper_request)
+            .unwrap();
+        let outgoing = store_context
+            .data_mut()
+            .durable_ctx_mut()
+            .as_wasi_http_view()
+            .new_response_outparam(sender)
+            .unwrap();
+
+        // unsafety comes from scope_and_collect:
+        //
+        // This function is not completely safe:
+        // please see cancellation_soundness in [tests.rs](https://github.com/rmanoka/async-scoped/blob/master/src/tests.rs) for a test-case that suggests how this can lead to invalid memory access if not dealt with care.
+        // The caller must ensure that the lifetime ’a is valid until the returned future is fully driven. Dropping the future is okay, but blocks the current thread until all spawned futures complete.
+        unsafe {
+            async_scoped::TokioScope::scope_and_collect(|s| {
+                s.spawn(proxy.wasi_http_incoming_handler().call_handle(
+                    store_context,
+                    incoming,
+                    outgoing,
+                ));
+            })
+            .await
+        }
+    };
+
+    let res_or_error = match receiver.await {
+        Ok(Ok(resp)) => {
+            Ok(virtual_export_compat::http_incoming_handler::http_response_to_output(resp).await?)
+        }
+        Ok(Err(e)) => Err(anyhow::Error::from(e)),
+        Err(_) => {
+            // An error in the receiver (`RecvError`) only indicates that the
+            // task exited before a response was sent (i.e., the sender was
+            // dropped); it does not describe the underlying cause of failure.
+            // Instead we retrieve and propagate the error from inside the task
+            // which should more clearly tell the user what went wrong. Note
+            // that we assume the task has already exited at this point so the
+            // `await` should resolve immediately.
+            let task_exit = task_exits.remove(0);
+            let e = match task_exit {
+                Ok(r) => r.expect_err("if the receiver has an error, the task must have failed"),
+                Err(_e) => anyhow!("failed joining wasm task"),
+            };
+            Err(e)?
+        }
+    };
+
+    let mut store_context = store.as_context_mut();
+
+    let current_fuel_level = store_context.get_fuel().unwrap_or(0);
```

Comment:
These  post-invocation steps are now duplicated, I guess, maybe worth extracting them to a function